### ISSUE TITLE
[Question]Significant differences between deepspeed and torchrun training results

### ISSUE THREAD

**Issue by getao** [microsoft/DeepSpeed#4762]
Hello,

I'm performing continual training of Llama-2-7b. I'm using the same hyperparameters to train the model (with flashattn2 and bf16 on A100) using torchrun and deepspeed (zero2) but the results are very different.

Torchrun's results are significantly better than deepspeed trained ones:

<img width="1778" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/12735658/dfb7a41a-78de-448e-bb79-5136f734670b">

My deepspeed configuration is:
```
{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1
    },
    "bf16": {
           "enabled": "auto"
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },

    "scheduler": {
        "type": "WarmupDecayLR",
        "params": {
            "warmup_min_lr": 5e-6,
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto",
            "warmup_type": "linear",
            "total_num_steps": "auto"
        }
    },

    "zero_optimization": {
        "stage": 2,
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
        "allgather_partitions": true,
        "allgather_bucket_size": 2e8,
        "overlap_comm": true,
        "reduce_scatter": true,
        "reduce_bucket_size": 2e8,
        "contiguous_gradients": true
    },

    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto"
}
```

I checked the differences between torchrun and deepspeed. The only possible difference is on the warmup_min_lr (torchrun using 0 but deepspeed using 5e-6) and optimizer (torchrun using adamw_torch, deepspeed using its native AdamW). Except these differences, I can't find any other possible reasons accounting for the difference of the performance. However, I don't think these two configuration differences could lead to such a significant performance gap. Is there anything that I missed? Any suggestion or comment would be very helpful to me.

Thank you

**Comment by getao** [comment]
After trying a group of experiments, I find the difference is mainly caused by "offload_optimizer". When I remove offloading, the results become much better and more similar to the torchrun's results.

Does it make sense?

**Comment by tjruwase** [comment]
@getao, thanks for the update. This suggests a bug in the optimizer offloading logic. The model performance gap is very significant, and we need to investigate. Are you able to provide more details to help us repro?

**Comment by getao** [comment]
Sure.

I continuously train the Llama-2-7b on a domain-specific dataset using 4 80GB A100 GPU. The training scripts of torchrun and deepspeed are:

```
#torchrun version
#!/bin/bash

lr=6e-5
eval_steps=1000
save_steps=2500
INPUT_DATA=MY_DATA_PATH
batch=1
epoch=2
OUT_DIR=MY_OUTPUT_PATH
max_seq_len=1024

torchrun --nproc-per-node=4 baseline.py --adam_beta2 0.95 --adam_epsilon 1e-6 --num_train_epochs $epoch --per_device_train_batch_size $batch --per_device_eval_batch_size $batch --gradient_accumulation_steps 1 --add_special_tokens \
          --learning_rate $lr --logging_steps 10 --save_strategy 'no' --evaluation_strategy 'steps' --eval_steps $eval_steps --save_steps $save_steps --bf16 --max_seq_length $max_seq_len --output_dir $OUT_DIR --logging_dir $OUT_DIR/logs --data_path $INPUT_DATA | tee $OUT_DIR/train.log
```

```
#deepspeed version
#!/bin/bash

lr=6e-5
# train_steps=50
eval_steps=1000
save_steps=2500
INPUT_DATA=MY_INPUT_PATH
batch=1
epoch=2

ds_zero=1
max_seq_len=1024
ds_config=ds_zero${ds_zero}.json

OUT_DIR=$MY_OUTPUT_PATH


deepspeed --num_gpus=4 baseline.py --adam_beta2 0.95 --adam_epsilon 1e-6 --num_train_epochs $epoch --per_device_train_batch_size $batch --per_device_eval_batch_size $batch --gradient_accumulation_steps 1 --add_special_tokens \
          --learning_rate $lr --logging_steps 10 --save_strategy 'no' --evaluation_strategy 'steps' --eval_steps $eval_steps --save_steps $save_steps --bf16 --max_seq_length $max_seq_len --output_dir $OUT_DIR --logging_dir $OUT_DIR --data_path $INPUT_DATA --deepspeed $ds_config | tee $OUT_DIR/train.log
```

Deepspeed configuration (no offloading version) is:
```
{
    "fp16": {
        "enabled": "auto",
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1
    },
    "bf16": {
           "enabled": "auto"
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "torch_adam": true,
            "lr": "auto",
            "betas": "auto",
            "eps": "auto",
            "weight_decay": "auto"
        }
    },

    "scheduler": {
        "type": "WarmupDecayLR",
        "params": {
            "warmup_min_lr": "auto",
            "warmup_max_lr": "auto",
            "warmup_num_steps": "auto",
            "warmup_type": "linear",
            "total_num_steps": "auto"
        }
    },

    "zero_optimization": {
        "stage": 1,
        "allgather_partitions": true,
        "allgather_bucket_size": 2e8,
        "overlap_comm": true,
        "reduce_scatter": true,
        "reduce_bucket_size": 2e8,
        "contiguous_gradients": true
    },

    "gradient_accumulation_steps": "auto",
    "gradient_clipping": "auto",
    "train_batch_size": "auto",
    "train_micro_batch_size_per_gpu": "auto"
}
```

The offloading version additionally has 
```
        "offload_optimizer": {
            "device": "cpu",
            "pin_memory": true
        },
``` in the "zero_optimization".

**Comment by getao** [comment]
Just run a group of experiments using the same version (0.12.3), same hyperparameters and same seed. ZERO-2 with offloading yields a significantly worse result:

<img width="2304" alt="image" src="https://github.com/microsoft/DeepSpeed/assets/12735658/f205fb2a-7bd8-49ab-a42f-d1616bd2a97d">

**Comment by jiqizaisikao** [comment]
I set some trainable parameters in my network and found that some layers update very slowly and hardly update at all. Some update very quickly. It is normal to update without using deepspeed.

**Comment by MaddyThakker** [comment]
Any updates on this? Were you able to find the reason for this?
